nwp/cams/_factories.py

```diff
@@ -0,0 +1,230 @@
+    contentLength: int
+    location: str
+    var: str
+    inittime: dt.datetime
+
+
+def make_asset_definitions(
+    opts: MakeAssetDefinitionsOptions,
+) -> tuple[dg.AssetsDefinition, dg.AssetsDefinition]:
+    """Generate the assets for a CAMS datset."""
+
+    @dg.asset(
+        name="source_archive",
+        key_prefix=opts.key_prefix(),
+        partitions_def=opts.partitions,
+        compute_kind="network_request",
+        op_tags={
+            "expected_runtime": "20min",
+            "MAX_RUNTIME_SECONDS_TAG": 20 * 60,
+        },
+    )
+    def _cams_source_archive(context: dg.AssetExecutionContext) -> list[CamsFileInfo]:
+        """Asset detailing all wanted remote files from CAMS."""
+        execution_start = dt.datetime.now(tz=dt.UTC)
+
+        it = map_partition_to_time(context)
+        # Check if partition is targeting a time more than 30 days old
+        # * CAMS data older than 30 days is only available from tape
+        # * These variables are slower to collect
+        use_slow: bool = False
+        if (dt.datetime.now(tz=dt.UTC) - it) > dt.timedelta(days=30):
+            use_slow = True
+
+        fis: list[CamsFileInfo] = []
+
+        # First handle single level variables
+        if opts.singlelevel_vars is not None:
+            for var in opts.singlelevel_vars.slow if use_slow else opts.singlelevel_vars.fast:
+                # Build the request
+                sl_var_request: dict[str, Any] = {
+                    "date": it.strftime("%Y-%m-%d/%Y-%m-%d"),
+                    "type": "forecast",
+                    "format": opts.file_format,
+                    "variable": var,
+                    "leadtime_hour": opts.singlelevel_vars.hours,
+                    "time": it.strftime("%H:%M"),
+                }
+                if opts.area == "eu":
+                    sl_var_request["model"] = "ensemble"
+
+                result = opts.client.retrieve(
+                    name=opts.dataset_name(),
+                    request=sl_var_request,
+                )
+                fis.append(CamsFileInfo(**result.toJSON(), var=var, inittime=it))
+
+        # Then handle multilevel variables
+        if opts.multilevel_vars is not None:
+            for var in opts.multilevel_vars.slow if use_slow else opts.multilevel_vars.fast:
+                # Build the request
+                ml_var_request: dict[str, Any] = {
+                    "date": it.strftime("%Y-%m-%d/%Y-%m-%d"),
+                    "type": "forecast",
+                    "format": opts.file_format,
+                    "variable": var,
+                    "leadtime_hour": opts.multilevel_vars.hours,
+                    "time": it.strftime("%H:%M"),
+                }
+                if opts.area == "eu":
+                    ml_var_request["model"] = "ensemble"
+                    ml_var_request["level"] = opts.multilevel_levels
+                else:
+                    # Jacob, why is this different for global?
```

Comment:
Ah, not sure why its different, never noticed that, I would think `level` should be the one for both, that's what is used for ERA5 as well and the like...